flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/catalog/CatalogCalciteSchema.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.catalog;
+
+import org.apache.flink.annotation.Internal;
+
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.rel.type.RelProtoDataType;
+import org.apache.calcite.schema.Function;
+import org.apache.calcite.schema.Schema;
+import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.schema.SchemaVersion;
+import org.apache.calcite.schema.Schemas;
+import org.apache.calcite.schema.Table;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * A mapping between Flink's catalog and Calcite's schema. This enables to look up and access objects(tables, views,
+ * functions, types) in SQL queries without registering them in advance. Databases are registered as sub-schemas
+ * in the schema.
+ */
+@Internal
+public class CatalogCalciteSchema implements Schema {
+
+	private final String catalogName;
+	private final Catalog catalog;
+
+	public CatalogCalciteSchema(String catalogName, Catalog catalog) {
+		this.catalogName = catalogName;
+		this.catalog = catalog;
+	}
+
+	/**
+	 * Look up a sub-schema (database) by the given sub-schema name.
+	 *
+	 * @param schemaName name of sub-schema to look up
+	 * @return the sub-schema with a given database name, or null
+	 */
+	@Override
+	public Schema getSubSchema(String schemaName) {
+
```

Comment:
This name `getSubSchema` seems a little misleading, since we always returns a `DatabaseCalciteSchema`, i know that `getSubSchema` is a interface defined by Calcite, but maybe we can have an abstraction of our own.

Reply:
do you mean we need an interface which defines methods like: `getDatabase(String database)` or `getTable(String table)` ?

Reply:
I don't necessarily understand @danny0405's concerns. We do have abstractions of our own. This class is just an adapter between our abstractions (Catalog, CatalogManager, Database etc.) and Calcite's (Schema).

Reply:
@dawidwys If we see Calcite's `Schema` interfaces definition, `Schema#getTable`, `Schema#getTableNames`, `Schema#getFunction` ... It means that for every level schema, we could fetch table/functions, but for this abstraction, `CatalogCalciteSchema` can only see database, and `DataBaseCalciteSchema` can only see table.

Reply:
Still don't necessarily get that. What do you suggest? To have additional level of abstractions:

    `Schema` <- `SubSchemaOnlySchema` <- `CatalogCalciteSchema` 

where `DatabaseOnlySchema` would have methods like that:

	@Override
	public Schema getSubSchema(String schemaName) {
		getDatabaseSchema(schemaName)
	}

        protected DatabaseCalciteSchema getDatabaseSchema(String schemaName)

or sth similar? I think this is would be over-engineering, but I am not strongly against it

Reply:
That makes the schema more complex, i think.